### PR TITLE
[SPARK-44475][SQL][CONNECT] Relocate DataType and Parser to sql/api

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

This PR relocates Parser and DataType family to sql/api.

### Why are the changes needed?

To extract DataType and Parser as an API into a shared module.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 1]
```diff
@@ -16,34 +16,20 @@
...
-  val OCTAL_CHAR_PATTERN = """\\([01][0-7]{2})(?s).*""".r
-  val ESCAPED_CHAR_PATTERN = """\\((?s).)(?s).*""".r
-
-  /** Get the command which created the token. */
-  def command(ctx: ParserRuleContext): String = {
-    val stream = ctx.getStart.getInputStream
-    stream.getText(Interval.of(0, stream.size() - 1))
-  }
-
+object ParserUtils extends SparkParserUtils {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:32]: Can't you move the entire utils class?

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkAnalysisUtils.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.util
+
+import org.apache.spark.sql.errors.DataTypeErrors
+
+trait SparkAnalysisUtils {
```
Comment (hvanhovell) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkAnalysisUtils.scala:21]: AttributeNameParser?

File: sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 3]
```diff
@@ -143,7 +142,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       // the scale is 2. The expected precision should be 2.
       this._precision = decimal.scale
       this._scale = decimal.scale
-    } else if (decimal.scale < 0 && !SQLConf.get.allowNegativeScaleOfDecimalEnabled) {
+    } else if (decimal.scale < 0 && !SqlApiConf.get.allowNegativeScaleOfDecimalEnabled) {
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:145]: how do we bind `SqlApiConf.get` to the active SQLConf at the server side?

Comment (hvanhovell) [reply sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:145]: As soon as SQLConf is loaded we connect them: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L185

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:32]: We could. I happed to move the least part that needed in this refactoring PR and it seems to be enough for moving a small portion.

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkAnalysisUtils.scala (hunk 2 above)
Comment (amaliujia) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkAnalysisUtils.scala]: done

**Comment by cloud-fan** [comment]
The k8s failure is unrelated. Thanks, merging to master/3.5 (for scala client)!

**Review by JackBuggins — COMMENTED** [review]
@amaliujia @dongjoon-hyun, Apologies if I am missing something or have this wrong, but I think there can be breaking user facing changes if an API has some function expecting a StructType param, then they rely on being able to call toAttributes based on what it is passed based on that param. 

I appreciate the objective of the change, and perhaps there is a nice way to handle this now when we are provided with a StructType obj and we want to use this from it, however StructType is annotated as Stable so I didn't expect changes to come here.

Could we either provide an example of what best to do in that situation either here or in the docs when users hit this scenario to help mitigate any challenges that will pose to users, or otherwise consider keeping an implementation, even marked as deprecated in StructType that makes use of the implementation from DataTypeUtils?

**Comment by JackBuggins** [comment]
Is it as simple as `DataTypeUtils.toAttributes(schema)`, where schema is some StructType?

**Comment by amaliujia** [comment]
Yes it is as simple as `DataTypeUtils.toAttributes(schema)` and same for `fromAttributes`.

**Comment by JackBuggins** [comment]
Thank you @amaliujia. As for StructType's interface stability and the removal, do you have an opinion on StructType maintaining an implementation for the above type of case?

**Comment by hvanhovell** [comment]
@JackBuggins while `StructType` is definitely stable & public API, the `toAttributes` methods was not. It was clearly marked as `protected[sql]` and it returned a sequence of an internal catalyst type (`Attribute`). We can always change those methods.

There is no real way to bring back this particular method. We can however provide an implicit conversion that will make life a bit easier, e.g.:
```scala
implicit class RichStructType(val schema: StructType) extends AnyVal {
  def toAttributes: Seq[Attribute] = DataTypeUtils.toAttributes(schema)
}

```

**Pull request closed without merging**